language/move-vm/integration-tests/src/tests/bad_storage_tests.rs

```diff
@@ -0,0 +1,159 @@
+use crate::compiler::compile_units;
+use move_core_types::{
+    account_address::AccountAddress,
+    gas_schedule::{GasAlgebra, GasUnits},
+    identifier::Identifier,
+    language_storage::StructTag,
+    vm_status::StatusType,
+};
+use move_lang::compiled_unit::CompiledUnit;
+use move_vm_runtime::move_vm::MoveVM;
+use move_vm_test_utils::{convert_txn_effects_to_move_changeset_and_events, InMemoryStorage};
+use move_vm_types::gas_schedule::{zero_cost_schedule, CostStrategy};
+use vm::file_format::{CompiledModule, CompiledScript};
+
+fn as_module(unit: CompiledUnit) -> CompiledModule {
+    match unit {
+        CompiledUnit::Module { module, .. } => module,
+        CompiledUnit::Script { .. } => panic!("expected module got script"),
+    }
+}
+
+fn as_script(unit: CompiledUnit) -> CompiledScript {
+    match unit {
+        CompiledUnit::Module { .. } => panic!("expected script got module"),
+        CompiledUnit::Script { script, .. } => script,
+    }
+}
+
+const TEST_ADDR: AccountAddress = AccountAddress::new([42; AccountAddress::LENGTH]);
+
+#[test]
+fn test_malformed_resource() {
+    let code = r#"
+        address 0x1 {
+            module Signer {
+                native public fun borrow_address(s: &signer): &address;
+
+                public fun address_of(s: &signer): address {
+                    *borrow_address(s)
+                }
+            }
+        }
+
+        module M {
+            use 0x1::Signer;
+
+            resource struct Foo { x: u64, y: bool }
+
+            public fun publish(s: &signer) {
+                move_to(s, Foo { x: 123, y : false });
+            }
+
+            public fun check(s: &signer) acquires Foo {
+                let foo = borrow_global<Foo>(Signer::address_of(s));
+                assert(foo.x == 123 && foo.y == false, 42);
+            }
+        }
+
+        script {
+            use {{ADDR}}::M;
+
+            fun main(s: &signer) {
+                M::publish(s);
+            }
+        }
+
+        script {
+            use {{ADDR}}::M;
+
+            fun main(s: &signer) {
+                M::check(s);
+            }
+        }
+    "#;
+    let code = code.replace("{{ADDR}}", &format!("0x{}", TEST_ADDR.to_string()));
+    let mut units = compile_units(TEST_ADDR, &code).unwrap();
+
+    let s2 = as_script(units.pop().unwrap());
```

Comment:
As we already about this offline I tried using expect  instead of unwrap and it ended up being a lot more verbose and didn't help clarify the intention here. I added comments explaining what each step is doing instead.